### ISSUE TITLE
Potential bug in prompt tuning

### ISSUE THREAD

**Issue by mayank31398** [huggingface/peft#54]
If yes, can you add an example?
I currently see attention masks like this: [1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1]
@pacman100

**Comment by mayank31398** [comment]
Also, I think this is also going to mess up the position ids in the model. Especially alibi stuff in BLOOM

**Comment by pacman100** [comment]
Hello @mayank31398, as mentioned in point 3 of https://github.com/huggingface/peft#caveats, `generate`  is not supported for Prompt Tuning/P-Tuning.

```
P_TUNING or PROMPT_TUNING doesn't support generate functionality of transformers bcause generate strictly requires input_ids/decoder_input_ids but P_TUNING/PROMPT_TUNING appends soft prompt embeddings to input_embeds to create new input_embeds to be given to the model. Therefore, generate doesn't support this yet.
```

**Comment by pacman100** [comment]
It would be great if you could provide a minimal reproducible example for the bug in training for the Prompt Tuning for the CLM models

**Comment by mayank31398** [comment]
@pacman100 but I see the method generate implemented in `PeftModelForCausalLM`. Am I missing something?

**Comment by mayank31398** [comment]
This problem will only occurs if padding is on left side. I think right side padded models should be fine

**Comment by pacman100** [comment]
Hello @mayank31398, I don't think it would be a problem as those pad tokens are unattended and have `0` in the attention mask. With position embeddings:
1. for models employing ALiBi, the relative positions of `<prompt>` and `<input_text>` is maintained just that the weightage is reduced as the distance is increased due to intermediate `<prompt>` tokens.
2. for models employing sinusoidal or frozen learned encodings, they are just lookup tables. While pre-training, no `<pad>` tokens are observed due to fully packing the context window, so adding `<pad>` tokens itself to the left side should already be leading to different position embeddings for input ids as they are now shifted by certain lengths to the right.  So, this is unexplored territory.
3. I did a quick visualization of attention masks of `prefix-tuning` which is highly performant and has huge performance gains over `prompt-tuning`. `prefix-tuning` incorporates separate prompt tokens to each layer unlike `prompt-tuning` which only incorporates it at the start. It also supports `generate` method. Below screenshot shows that during inference it is able to attend to the virtual token embeddings properly.

![Screenshot 2023-02-06 at 11 11 08 PM](https://user-images.githubusercontent.com/13534540/217045169-5b3d157b-57ea-499b-8638-fcec700f79ee.png)

**Comment by mayank31398** [comment]
Um, but its still not correct right?
For models that have learnt positional embeddings/ sinusoidal?
For instance, the positional embeddings assigned in the case 110011 vs 001111
pos embedding is different for the last 2 tokens in both cases.
in the first case, it gets pos_5 and pos_6 and in the second, it gets pos_3, pos_4

**Comment by mayank31398** [comment]
But is this also done during training?
I only see it being done during generation. `prepare_inputs_for_generation`

**Comment by mayank31398** [comment]
```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from ..peft import get_peft_model, TaskType
from ..peft.tuners.prompt_tuning import PromptTuningConfig, PromptTuningInit

model_name = "bigscience/bloom-560m"

model = AutoModelForCausalLM.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name)

peft_model = get_peft_model(
    model, 
    PromptTuningConfig(
        task_type=TaskType.CAUSAL_LM,
        prompt_tuning_init=PromptTuningInit.TEXT,
        num_virtual_tokens=8,
        prompt_tuning_init_text="Classify the sentiment of the sentence:",
        tokenizer_name_or_path=model_name,
    )
)

# no padding
def construct_vanilla_inputs(prompt, input, output):
    x = tokenizer(prompt + input)
    input_ids = x["input_ids"]
    attention_mask = x["attention_mask"]
    labels = tokenizer(output)["input_ids"]

    in_len = len(input_ids)

    input_ids = input_ids + labels
    attention_mask = attention_mask + [1] * len(labels)
    labels = [-100] * in_len + labels
    # labels = input_ids

    return {
        "input_ids": input_ids,
        "attention_mask": attention_mask,
        "labels": labels
    }


# left padding
def modify1(input_ids, attention_mask, labels, n=100):
    input_ids = list(range(n)) + input_ids
    attention_mask = [0] * n + attention_mask
    labels = [-100] * n + labels

    return {
        "input_ids": input_ids,
        "attention_mask": attention_mask,
        "labels": labels
    }


# middle padding
def modify2(input_ids, attention_mask, labels, n=100):
    input_ids = input_ids[:5] + list(range(n)) + input_ids[5:]
    attention_mask = attention_mask[:5] + [0] * n + attention_mask[5:]
    labels = labels[:5] + [-100] * n + labels[5:]

    return {
        "input_ids": input_ids,
        "attention_mask": attention_mask,
        "labels": labels
    }


# right padding
def modify3(input_ids, attention_mask, labels, n=100):
    input_ids = input_ids + list(range(n))
    attention_mask = attention_mask + [0] * n
    labels = labels + [-100] * n

    return {
        "input_ids": input_ids,
        "attention_mask": attention_mask,
        "labels": labels
    }


# peft test
def construct_peft_inputs(input, output):
    x = tokenizer(input)
    input_ids = x["input_ids"]
    attention_mask = x["attention_mask"]
    labels = tokenizer(output)["input_ids"]

    in_len = len(input_ids)

    input_ids = input_ids + labels
    attention_mask = attention_mask + [1] * len(labels)
    labels = [-100] * in_len + labels

    return {
        "input_ids": input_ids,
        "attention_mask": attention_mask,
        "labels": labels
    }


def f(x: dict):
    for i in x:
        x[i] = torch.tensor([x[i]])
    return x


prompt = "Classify the sentiment of the sentence:"
input = " The movie sucks\nSentiment:"
output = " negative</s>"

loss = model(**f(construct_vanilla_inputs(prompt, input, output))).loss
print("no pad")
print(loss)

loss = model(**f(modify1(**construct_vanilla_inputs(prompt, input, output)))).loss
print("left pad")
print(loss)

loss = model(**f(modify2(**construct_vanilla_inputs(prompt, input, output)))).loss
print("middle pad")
print(loss)

loss = model(**f(modify3(**construct_vanilla_inputs(prompt, input, output)))).loss
print("right pad")
print(loss)

loss = peft_model(**f(construct_peft_inputs(input, output))).loss
print("peft")
print(loss)
```

The output is:
```shell
no pad
tensor(6.1106)
left pad
tensor(6.1107)
middle pad
tensor(6.1107)
right pad
tensor(6.1106)
peft
tensor(6.1106, grad_fn=<NllLossBackward0>)
```

Looks correct to me :)

**Comment by mayank31398** [comment]
Still a minute difference in there. Would you suggest to ignore it?

**Comment by mayank31398** [comment]
Also, another question @pacman100:
you mentioned that generation currently doesn't work for prompt tuning.
are there plans to support it?

**Comment by mayank31398** [comment]
Also, feel free to close this :)

**Comment by mayank31398** [comment]
Let me know if I can add this as a unit test or something if you prefer :)

**Comment by pacman100** [comment]
Hello @mayank31398, thank you for the example script and as per it things look fine.

wrt `generate` functionality support, this commit should enable it 
https://github.com/huggingface/peft/pull/55/commits/bdeb3778d0a35d0fd66f87f884e6f8fe32ecf705

```
outputs = model.generate(input_ids=inputs["input_ids"], attention_mask=inputs["attention_mask"], max_new_tokens=10, eos_token_id=3)
```

I am seeing below result for one of the raft task using `bloom-560m`
![Screenshot 2023-02-07 at 3 17 48 PM](https://user-images.githubusercontent.com/13534540/217210136-acc8c8c7-2266-49c9-9924-dddd455e289c.png)

**Comment by pacman100** [comment]
All contributions are welcome wrt tests, features, documentation and examples 🤗🚀✨.
Please raise PRs with the tests that you have planned. 
Thank you for taking the time to try this repo. Keep the feedback incoming!

**Comment by mayank31398** [comment]
Thanks. Closing this issue
